Port to dune

Leonidas-from-XIV opened the pull request:
Given #54 has somewhat stalled, this is a PR to upstream the dune port that we successfully use in @dune-universe.

The @dune-universe ports generally try to touch as little code as possible but if this code end up upstream I think it makes sense to touch some of the OCaml code to remove some low-hanging fruits like some of the warnings that dune enables by default. I went ahead and fixed some but I'll gladly fix more.

kit-ty-kate commented on extlib.opam:
```diff
@@ -19,17 +19,11 @@ authors: [
-  "ocamlfind" {build}
   "cppo" {build}
-  "base-bytes" {build}
+  "base-bytes"
```
This is not required anymore since `dune` requires OCaml >= 4.02
```suggestion
```

kit-ty-kate commented on extlib.opam:
```diff
@@ -19,17 +19,11 @@ authors: [
-install: [ [make "minimal=1" "install"] ]
 depends: [
+  "dune"
   "ocaml"
```
```suggestion
  "ocaml" {>= "4.02"}
```

kit-ty-kate commented on extlib.opam:
```diff
@@ -19,17 +19,11 @@ authors: [
-]
-install: [ [make "minimal=1" "install"] ]
 depends: [
+  "dune"
```
```suggestion
  "dune" {>= "1.0"}
```

Leonidas-from-XIV commented on src/configure.ml:
```diff
@@ -1,18 +1,17 @@
+    print_endline "WITH_BYTES";
     exit 0
   | [|_;"-compile-args"|] ->
     if Sys.ocaml_version >= "4.00.0" then print_endline "-bin-annot";
-    show_bytes "-package bytes";
+    print_endline "-package bytes";
     exit 0
```
No, not if building with dune on 4.02+ which does not need `bytes`, `base-bytes` or `ocamlfind`.

If we decide to drop the Makefile support, then we can delete the whole `configure.ml` (and maybe have a `Makefile` that just calls dune as a courtesy to users used to calling `make`)

ygrek replied:
yes, can drop support for building without bytes

Leonidas-from-XIV replied:
I've dropped it.

Leonidas-from-XIV commented on test/std.ml:
```diff
@@ -1,24 +0,0 @@
-(* check compatibility of interfaces *)
```
You're right. It created issues because it is named the same as `Std` in extlib so the tests would accidentally pick this up, so I removed it but actually renaming is enough to work around that problem.

Et7f3 commented on test/dune:
```diff
@@ -0,0 +1,9 @@
+(executable
+  (name runner)
+  (flags :standard -w -3)
+  (modules :standard \ test_Unzip test_UTF8 test_Base64)
```
They aren't tested now ?

kit-ty-kate replied:
They haven’t been included in the standard distributions for quite a few versions already:
* `Base64` since 1.7.8 (released last year)
* `Unzip`, `UTF8` and `UChar` since 1.6.0 (released in 2013)

Leonidas-from-XIV replied:
Given they have not been part of the distribution (since everyone installs whatever OPAM ships with which is the "minimal distribution" which does not include these modules) I would suggest just deleting these modules and their tests. This makes the distribution simpler, the tests simpler and makes no difference for people using 1.7.8+.

ygrek commented:
> Btw i think it would be nice to remove the now always true #if OCAML >= 400 and so on for all versions < 4.02

yes, but not in this PR please

ygrek commented:
oh, ic, too late, I liked how it was a small contained PR :cry:

kit-ty-kate commented:
> oh, ic, too late, I liked how it was a small contained PR 😢

I’m assuming @Leonidas-from-XIV is off work at this time so I’ve gone ahead and opened https://github.com/ygrek/ocaml-extlib/pull/64 for you. I’ve done my best to make it as small and contained as possible, I hope it’s ok.

ygrek commented:
Thanks, merged the minimal PR, will look to pick other changes

Leonidas-from-XIV commented:
Thanks @kit-ty-kate for extracting the minimal version again, I've rebased on master so this PR is about simplifying the leftovers.

@ygrek OPAM vendors Extlib so I talked to them. They already used a custom dune file for Extlib, so they said porting to dune will make it actually easier for them, since they don't have to maintain their own version anymore.

Leonidas-from-XIV commented:
Okay, that took a bit. `test_std.ml` is somewhat tricky because newer versions of OCaml added quite some functions, so I exposed these functions in those versions as well as manually overriding those that have different signatures than what OCaml ships with (but at least we are aware of these functions). The downside is that the test is fragile since it depends on the signatures of Stdlib modules. We could consider including the original signature in the `ext<Module>.mli` files, but for now I've just added the signatures by hand in version gated comments.

I also removed the `configure.ml` (which required a bit of an ugly hack because apparently CPPO can't handle `#if FOO` when run with `-D "FOO true"` — looks like its support for booleans is a bit wonky) and `install.ml` since all the configuration options can now be directly passed from `dune` to `cppo`)

ygrek commented:
please keep configure.ml, I want to keep make build for a while (maybe echo deprecation message), there is probably not many non-opam users left, but it doesn't cost much to keep compatibility for some time

Leonidas-from-XIV commented:
I've added `configure.ml` back. I had to adjust it slightly since the options to CPPO changed to match the `dune` file so the source code would work with both.

Leonidas-from-XIV commented:
I've rebased on master and since the `test_std.ml`/`check_stdlib,ml` was fixed to be compatible with all supported versions of OCaml I added it back into the test suite. That said, it is fragile and has to be adapted to each release of OCaml. But it shows pretty well the differences between different OCaml versions and how to deal with them.

ygrek commented:
> it is fragile and has to be adapted to each release of OCaml. But it shows pretty well the differences between different OCaml versions and how to deal with them.

it is precisely the goal, and also that's why I am not sure it has to be part of test suite. It is ok to make quick extlib release to stay compatible with new ocaml version while not having back-ported all the missing functions etc. Ie. my workflow is to run it manually and fix when I have time, but tests suite is smth that has to be green always otherwise it is grave bug.

Leonidas-from-XIV commented:
Well, fair enough, moving it out is pretty easy and I see your point.

ygrek commented:
i forgot about this and cleaned up some things in https://github.com/ygrek/ocaml-extlib/commit/3d9caa998545909a412a6112c3ef899b63224211
so this PR now conflicts
I do not plan to drop Makefile yet
are there any changes worth recovering from this PR left?
